packages/nimble-components/src/table/index.ts

```diff
@@ -259,45 +298,105 @@ export class Table<
-        this.table.setOptions(prev => ({
-            ...prev,
-            ...this.options,
-            state,
-            onStateChange: (updater: unknown) => {
-                const updatedState = typeof updater === 'function'
-                    ? (updater(state) as TanStackTableState)
-                    : (updater as TanStackTableState);
-                this.update(updatedState);
+    private setSortState(): void {
+        const sortedColumns = this.getColumnsParticipatingInSorting().sort(
+            (x, y) => x.sortIndex! - y.sortIndex!
+        );
+        this.firstSortedColumn = sortedColumns.length
+            ? sortedColumns[0]
+            : undefined;
+
+        const tanStackSortingState: TanStackSortingState = sortedColumns.map(
+            column => {
+                return {
+                    id: column.internalUniqueId,
+                    desc:
+                        column.sortDirection
+                        === TableColumnSortDirection.descending
+                };
             }
-        }));
-    };
+        );
 
-    // Generate columns for TanStack that correspond to all the keys in TData because all operations,
-    // such as grouping and sorting, will be performed on the data's records, not the values rendered within a cell.
-    private generateTanStackColumns(data: readonly TData[]): void {
-        if (data.length === 0) {
-            return;
-        }
+        this.updateTableOptions({
+            state: {
+                sorting: tanStackSortingState
+            }
+        });
+    }
 
-        const firstItem = data[0]!;
-        const keys = Object.keys(firstItem);
-        const generatedColumns = keys.map(key => {
+    private generateTanStackColumns(): void {
+        const generatedColumns = this.columns.map(column => {
             const columnDef: TanStackColumnDef<TData> = {
-                id: key,
-                accessorKey: key,
-                header: key
+                id: column.internalUniqueId,
+                accessorKey: column.operandDataRecordFieldName,
+                sortingFn: this.getTanStackSortingFunction(column.sortOperation)
             };
             return columnDef;
         });
 
-        this.updateTableOptions({ columns: generatedColumns });
+        this.updateTableOptions({
+            // Force TanStack to detect a data update because a columns's accessor is
+            // referenced when creating a new row model.
+            data: this.copyData(this.table.options.data),
+            columns: generatedColumns
+        });
     }
+
+    private copyData(data: readonly TData[]): TData[] {
+        return data.map(record => {
+            return { ...record };
+        });
+    }
+
+    private getTanStackSortingFunction(
+        sortOperation: TableColumnSortOperation
+    ): TanStackSortingFnOption<TData> {
+        switch (sortOperation) {
+            case TableColumnSortOperation.basic:
+                return 'basic';
```

Comment:
Can we return the TanStack `basic` sorting function itself, so we don't have to use strings?

Reply:
Ah, then I don't feel strongly about this at all. The main concern on my mind was to make sure we would break appropriately should TanStack change its API, which it seems like we will.

Reply:
I think the nice thing about the new approach is that it makes the code very explicit that we are using an algorithm coming from TanStack. So, unless there is opposing feedback, I'll leave it the way I've updated it.